clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -303,53 +303,60 @@ public boolean matchesSubscribedPattern(String topic) {
      * can do the partition assignment (which requires at least partition counts for all topics
      * to be assigned).
      *
-     * Note this is thread-safe since the Set is backed by a ConcurrentMap.
-     *
      * @return The union of all subscribed topics in the group if this member is the leader
      *   of the current generation; otherwise it returns the same set as {@link #subscription()}
      */
-    public Set<String> groupSubscription() {
+    synchronized Set<String> groupSubscription() {
         return this.groupSubscription;
     }
 
-    /**
-     * Note this is thread-safe since the Set is backed by a ConcurrentMap.
-     */
-    public boolean isGroupSubscribed(String topic) {
+    synchronized boolean isGroupSubscribed(String topic) {
         return groupSubscription.contains(topic);
     }
 
-    private TopicPartitionState assignedState(TopicPartition tp) {
+    private synchronized TopicPartitionState assignedState(TopicPartition tp) {
         TopicPartitionState state = this.assignment.stateValue(tp);
         if (state == null)
             throw new IllegalStateException("No current assignment for partition " + tp);
         return state;
     }
 
+    private synchronized TopicPartitionState assignedStateOrNull(TopicPartition tp) {
+        return this.assignment.stateValue(tp);
+    }
+
     public void seek(TopicPartition tp, long offset) {
         assignedState(tp).seek(offset);
     }
 
+    void maybeSeek(TopicPartition tp, long offset, OffsetResetStrategy requestedResetStrategy) {
+        if (!assignedState(tp).maybeSeek(offset, requestedResetStrategy))
+            log.debug("Skipping reset of partition {} to offset {} " +
+                    "since it is no longer needed or an alternative reset has been requested", tp, offset);
+        else
+            log.info("Resetting offset for partition {} to offset {}.", tp, offset);
+    }
+
     /**
      * @return an unmodifiable view of the currently assigned partitions
      */
-    public Set<TopicPartition> assignedPartitions() {
+    public synchronized Set<TopicPartition> assignedPartitions() {
```

Comment:
This method is also not safe since we are returning a view to the underlying collection. As far as I can tell, we have only two uses for this method: 1) create a copy of the collection, 2) iterate over the members. Perhaps we can let this method do the copy so that it is protected with the lock to handle case 1. For case 2, potentially we can add a `forEachAssigned` or something like that which is also protected by the lock. What do you think?

Reply:
(1) seems fair.
(2) seems fair too. First I was debating whether we need that or not but for iterations it might be better to acquire the lock one time and not _n_ times given that the operations done in `TopicPartitionState` are fairly cheap too.

Reply:
What's even more I think for (2) it'd be enough to modify (or overload) requestOffsetReset to have list parameters so you can pass in multiple partitions.
(`subscriptions.assignedPartitions()` is only called from seekTo... methods which in turn call `requestOffsetReset`)

Reply:
Makes sense.